### DISCUSSION TITLE
Disassembling access to array of structures

### DISCUSSION THREAD

**Discussion by exander77** [NationalSecurityAgency/ghidra#5150] [Q&A]
How can I turn this:
```c
struct {
  uint8_t temp_in;
  uint8_t temp_out;
  uint16_t pwm;
}
```
From:
```c
      iVar1 = (uint)fan_tbl_p * 4;
      if ((&fan_pwm_row_ARRAY_0002be62[0].temp_in)[iVar1 + 4] < ec_temp) {
         enable_set_fanpwm((&fan_pwm_row_ARRAY_0002be62[0].temp_in)[iVar1 + 6]);
         fan_tbl_p = fan_tbl_p + 1;
      }
```
Into something like:
```c
      if (fan_pwm_row_ARRAY_0002be62[fan_tbl_p+1].temp_in < ec_temp) {
         enable_set_fanpwm(fan_pwm_row_ARRAY_0002be62[fan_tbl_p+1].pwm);
         fan_tbl_p = fan_tbl_p + 1;
      }
```

**Comment by Wall-AF** [comment]
Here's another example of bad array indexing or maybe the type definitions are awry (any suggestions?):
[dragon_writeComputationMap.zip](https://github.com/NationalSecurityAgency/ghidra/files/9739763/dragon_writeComputationMap.zip)

**Comment by exander77** [comment]
> Here's another example of bad array indexing or maybe the type definitions are awry (any suggestions?): [dragon_writeComputationMap.zip](https://github.com/NationalSecurityAgency/ghidra/files/9739763/dragon_writeComputationMap.zip)

It ain't perfect in this, although I would think that it isn't really hard to detect an array of structures. In my example above, it is really obvious that it is an array of structures.

**Comment by 0xBEEEF** [comment]
I have similar problems with arrays and structures here. But for me the problem is even deeper, because my structures contain fixed arrays of other structures, e.g. something like this:

```
struct InnerTest {
  double a;
  double b;
}

struct Test {
  int a;
  struct InnerTest b[20];
  int c;
}
```

I get similar, and even worse readable results when decompiling.

Structures of the following kind work especially badly for me when they are looped.

```
struct NamedStruct {
  char a[64];
  int b;
  int c;
  float d;
}
```

Sometimes the string in the array is not recognized correctly, or the access is always to the first character. 

Sorry, that I don't have an executable example at hand. I got this info quickly from memory, so no guarantee that this is executable. Just wanted to add these additions here because I think it's very important.

**Comment by Wall-AF** [comment]
In this example [dragon_setNewParValue.zip](https://github.com/NationalSecurityAgency/ghidra/files/9741253/dragon_setNewParValue.zip), by using my mechanism for mixed size pointers (see [discussion #4469](https://github.com/NationalSecurityAgency/ghidra/discussions/4469)) for the return value of `findParName(...)` Ghidra is able to interpret the nested array correctly.  If I replace that type with the standard `DgnParDetails... *` the good decompiled interpretation changes from:
`pDVar2.__lp__->m_a101Ints_0x6[nParIdx] = iVar1;`
to
`   *(int *)((int)pDVar2 + nParIdx * 2 + 6) = iVar1;`

Sorry, I missed the `.__lp__` as I've modified the decompiled display to hide it!

**Comment by Wall-AF** [comment]
I have another example of what should be a simple loop, executing 32 times and modifying 2 embedded arrays of type `ulong` (32 bits).  The structure (so far) is
```
struct DgnAmbienceComponent_0x1a8_t { /* size 0x1a8 */
    int aLowerVals?_0x0[32]; /* 11b0:1439 */
    int m_nIncrement_0x40; /* 11b0:130d */
    undefined field2_0x42;
    undefined field3_0x43;
    undefined2 field4_0x44;
    int m_nNewIncrement_0x46; /* 11b0:12f5 */
    byte m_1st32ByteAray_0x48[32];
    byte m_2nd32ByteArray_0x68[32];
    byte field8_0x88[32];
    ulong m_1st32LongArray_0xa8[32]; /* =val */
    ulong m_2nd32LongArray_0x128[32]; /* =int((val+0.5)/8) */
};
```
Ghidra produces
```
   i = 0;
...
   do {
      npAmbComp->m_1st32LongArray_0xa8[0] = ULONG_1210_744a * *pInts1;
      npAmbComp->m_2nd32LongArray_0x128[0] =
           ULONG_1210_744e * (*(uint (*) [32])pInts2)[0] + 4 >> 3;
      pInts1 = pInts1 + 1;
            /* np... is increasing to goto next array element */
      npAmbComp = (DgnAmbienceComponent_0x1a8_t *)&npAmbComp->field_0x4;
      pInts2 = pInts2 + 1;
      i += 1;
   } while (i < 32);
```
instead of
```
   i = 0;
...
   do {
      npAmbComp->m_1st32LongArray_0xa8[i] = ULONG_1210_744a * pInts1[i];
      npAmbComp->m_2nd32LongArray_0x128[i] = ULONG_1210_744e * (pInts2)[i] + 4) >> 3;
      i += 1;
   } while (i < 32);
```

I've included the zipped output from the `Debug Function Decompilation` dropdown menu in the `Decompiler:`Panel [here](https://github.com/NationalSecurityAgency/ghidra/files/12195056/dragon_RecalcAmbienceData.zip)

**Comment by Wall-AF** [comment]
Here's another example:
```
   numItems$1 = local_6->numItems_0x0;
   nIdx = 0;
   pTmp = (Dgn1158_17de_0x20a_t *)local_6;
   if (numItems$1 != 0) {
      do {
         FUN_1150_1b1c(pTmp->field5_0x6[0],param_2,param_4);
         pTmp = (Dgn1158_17de_0x20a_t *)&pTmp->field_0x4;
         nIdx += 1;
      } while (nIdx < numItems$1);
   }
```
instead of
```
   nIdx = 0;
   if (local_6->numItems != 0) {
      do {
         FUN_1150_1b1c(local_6->field5_0x6[nIdx],param_2,param_4);
      } while (++nIdx < local_6->numItems);
   }
```
from
```
                             LAB_1150_187d                                   XREF[3]:     1150:1799(j), 1150:184f(j), 
                                                                                          1150:182a(j)  
   1150:187d c4 5e fc     016           LES        BX,SS:[BP + local_6]
   1150:1880 26 8b 07     016           MOV        AX,word ptr ES:[BX]
   1150:1883 89 46 fa     016           MOV        word ptr SS:[BP + local_8],AX
   1150:1886 33 f6        016           XOR        SI,SI
   1150:1888 8b 7e fc     016           MOV        DI,word ptr SS:[BP + local_6]
   1150:188b 3b 76 fa     016           CMP        SI,word ptr SS:[BP + local_8]
   1150:188e 73 1e        016           JNC        LAB_1150_18ae
                             LAB_1150_1890                                   XREF[1]:     1150:18ac(j)  
   1150:1890 ff 76 10     016           PUSH       word ptr SS:[BP + param_4]
   1150:1893 66 ff 76 0a  018           PUSH       dword ptr SS:[BP + param_2]
   1150:1897 8e 46 fe     01c           MOV        ES,word ptr SS:[BP + local_6+0x2]
   1150:189a 66 26 ff     01c           PUSH       dword ptr ES:[DI + 0x6]
             75 06
   1150:189f e8 7a 02     020           CALL       FUN_1150_1b1c                                    void FUN_1150_1b1c(Dgn1150_1ee5_
   1150:18a2 83 c4 0a     020           ADD        SP,0xa
   1150:18a5 83 c7 04     016           ADD        DI,0x4
   1150:18a8 46           016           INC        SI
   1150:18a9 3b 76 fa     016           CMP        SI,word ptr SS:[BP + local_8]
   1150:18ac 72 e2        016           JC         LAB_1150_1890
```
